### PR TITLE
[WIP] #4367 ORMultiValueDictionary DeltaGroup cast to ORSet

### PR DESCRIPTION
**Pull request by Arkatufus**

### CODE REVIEW

**Comment by Arkatufus** [comment]
The problem lies in ORMultiValueDictionary delta group propagation, the current code didn't serialize/deserialize delta groups correctly because of its tight coupling and dependency with ORDictionary code and the need to support generics.

File: src/contrib/cluster/Akka.DistributedData.Tests/ReplicatorSpecs.cs [hunk 1]
```diff
@@ -270,24 +269,29 @@ private async Task ORMultiValueDictionary_Should_Merge()
...
             {
                 AwaitAssert(() =>
                 {
                     // update it with a replication factor of two
-                    _replicator2.Tell(Dsl.Update(_keyJ, ORMultiValueDictionary<string, string>.Empty,
-                        _writeTwo, x => x.AddItem(Cluster.Cluster.Get(_sys2), "a", "A")));
+                    _replicator2.Tell(Dsl.Update(
+                        _keyJ, 
+                        ORMultiValueDictionary<string, string>.Empty,
+                        WriteLocal.Instance,
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData.Tests/ReplicatorSpecs.cs:283]: why did all of these get changed to `WriteLocal`?

File: src/contrib/cluster/Akka.DistributedData.Tests/ReplicatorSpecs.cs [hunk 2]
```diff
@@ -374,6 +392,101 @@ private async Task ORMultiValueDictionary_Should_Merge()
...
+        {
+            await InitCluster();
+            await UpdateORMultiValueDictionaryNodes();
+        }
+
+        private async Task UpdateORMultiValueDictionaryNodes()
+        {
+            async Task SetEntryCoreAync(string key, IImmutableSet<string> vals)
+            {
+                // Bug only appears when we use WithValueDeltas
```
Comment (Aaronontheweb) [inline src/contrib/cluster/Akka.DistributedData.Tests/ReplicatorSpecs.cs:409]: Good to know!

**Comment by Aaronontheweb** [comment]
Debugging the reproduction spec locally:

```
>	Akka.DistributedData.dll!Akka.DistributedData.ORMultiValueDictionary<string, string>.Merge(Akka.DistributedData.ORMultiValueDictionary<string, string> other) Line 194	C#
 	Akka.DistributedData.dll!Akka.DistributedData.ORMultiValueDictionary<string, string>.Merge(Akka.DistributedData.IReplicatedData other) Line 258	C#
 	Akka.DistributedData.dll!Akka.DistributedData.Internal.DataEnvelope.Merge(Akka.DistributedData.IReplicatedData otherData) Line 500	C#
 	Akka.DistributedData.dll!Akka.DistributedData.Internal.DataEnvelope.Merge(Akka.DistributedData.Internal.DataEnvelope other) Line 476	C#
```

image

Problem appears to be that this check fails: `_withValueDeltas == other._withValueDeltas`.

Issue might be that we miss that boolean during serialization. I'm going to double check.

**Comment by Aaronontheweb** [comment]
@Arkatufus came up with a simpler reproduction:

```csharp
/// <summary>
        /// Bug reproduction: https://github.com/akkadotnet/akka.net/issues/4367
        /// </summary>
        [Fact]
        public void Bugfix_4367_ORMultiValueDictionary_Deltas_must_merge_other_ORMultiValueDictionary()
        {
            var m1 = ORMultiValueDictionary<string, string>.EmptyWithValueDeltas
                .SetItems(_node1, "a", ImmutableHashSet.Create("A"))
                .SetItems(_node1, "b", ImmutableHashSet.Create("B1"));

            var m2 = ORMultiValueDictionary<string, string>.EmptyWithValueDeltas
                .SetItems(_node2, "c", ImmutableHashSet.Create("C"))
                .SetItems(_node2, "b", ImmutableHashSet.Create("B2"));

            // This is how deltas really get merged inside the replicator
            var dataEnvelope = new DataEnvelope(m1.Delta);
            if (dataEnvelope.Data is IReplicatedDelta withDelta)
            {
                dataEnvelope = dataEnvelope.WithData(withDelta.Zero.MergeDelta(withDelta));
            }

            // Bug: this is was an ORDictionary<string, ORSet<string>> under #4302
            var storedData = dataEnvelope.Data;

            // simulate merging an update
            var merged1 = (ORMultiValueDictionary<string, string>)m2.Merge(storedData);

            merged1.Entries["a"].Should().BeEquivalentTo("A");
            merged1.Entries["b"].Should().BeEquivalentTo("B1", "B2");
            merged1.Entries["c"].Should().BeEquivalentTo("C");
        }
```

**Comment by Aaronontheweb** [comment]
Looks like this bug has multiple causes - saw your note about the serialization. That's definitely still a problem too. Looking into it now.

**Comment by Aaronontheweb** [comment]
```csharp
[Fact()]
        public void ReplicatedDataSerializer_should_serialize_ORMultiDictionary_delta()
        {
            CheckSerialization(ORMultiValueDictionary<string, string>.Empty.Delta);
            CheckSerialization(ORMultiValueDictionary<string, string>.EmptyWithValueDeltas.Delta);
            CheckSerialization(ORMultiValueDictionary<string, string>.Empty.AddItem(_address1, "a", "A").Delta);
            CheckSerialization(ORMultiValueDictionary<string, string>.EmptyWithValueDeltas
                .AddItem(_address1, "a", "A1")
                .SetItems(_address2, "b", ImmutableHashSet.CreateRange(new[] { "B1", "B2", "B3" }))
                .AddItem(_address2, "a", "A2").Delta);

            var m1 = ORMultiValueDictionary<string, string>.Empty.AddItem(_address1, "a", "A1").AddItem(_address2, "a", "A2");
            var m2 = ORMultiValueDictionary<string, string>.Empty.SetItems(_address2, "b", ImmutableHashSet.CreateRange(new[] { "B1", "B2", "B3" }));
            CheckSameContent(m1.Merge(m2), m2.Merge(m1));
            CheckSerialization(m1.Merge(m2).Delta);
        }
```
Also fails - looks like we never properly checked `ORMultiDictionaryDelta` serialization.

**Comment by Aaronontheweb** [comment]
Based on some careful debugging, it looks like this is where the problem occurs:

https://github.com/akkadotnet/akka.net/blob/e69d65c9efcbe42ac4e7503824f26679e61b33ae/src/contrib/cluster/Akka.DistributedData/Serialization/ReplicatedDataSerializer.cs#L895-L900

The `x.Value` is the delta operation itself - not the underlying ORSet.

**Pull request closed without merging**